File: Credicrew/backend/app/services/forecast.py
"""Forecast Studio engine — Python mirror of frontend/src/lib/forecast.ts.

Takes the current funnel + conversion + velocity assumptions and runs a
Monte-Carlo simulation of every candidate's forward walk through the
funnel. Returns the probability of a hire by the target date, the
earliest-hire-date distribution, the bottleneck stage, a sensitivity
tornado, and concrete recommendations.

Pure functions. Deterministic for a given seed so the API returns
byte-identical numbers across calls with the same input. The same
algorithm runs in the browser, so any programmatic / agentic client sees
the same forecast the recruiter is staring at on the page.
"""
from __future__ import annotations

import math
import time
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional


PROGRESSION = ["new", "outreach", "screening", "interview", "offer"]

STAGE_LABEL = {
    "new": "New",
    "outreach": "Outreach",
    "screening": "Screening",
    "interview": "Interview",
    "offer": "Offer",
}

DEFAULT_CONVERSION = {
    "new": 0.65,
    "outreach": 0.40,
    "screening": 0.60,
    "interview": 0.35,
    "offer": 0.70,
}

DEFAULT_VELOCITY = {
    "new": 2.0,
    "outreach": 4.0,
    "screening": 5.0,
    "interview": 8.0,
    "offer": 4.0,
}

DEFAULT_NOTICE_DAYS = 30
DEFAULT_DURATION_SIGMA = 0.45
DAY_MS = 86_400_000


# ---------- input dataclasses ----------


@dataclass
class ForecastAssumptions:
    conversion: dict[str, float] = field(default_factory=lambda: dict(DEFAULT_CONVERSION))
    velocity: dict[str, float] = field(default_factory=lambda: dict(DEFAULT_VELOCITY))
    notice_period_days: int = DEFAULT_NOTICE_DAYS
    duration_sigma: float = DEFAULT_DURATION_SIGMA


@dataclass
class ForecastInput:
    funnel: dict[str, int]
    target_date: str  # YYYY-MM-DD
    now: Optional[int] = None  # epoch ms
    assumptions: Optional[ForecastAssumptions] = None
    trials: int = 4000
    seed: Optional[int] = None


# ---------- RNG (mulberry32, matching the TS engine bit-for-bit) ----------


def _mulberry32(seed: int):
    state = [seed & 0xFFFFFFFF]

    def rng() -> float:
        state[0] = (state[0] + 0x6D2B79F5) & 0xFFFFFFFF
        t = state[0]
        t = ((t ^ (t >> 15)) * (t | 1)) & 0xFFFFFFFF
        t ^= (t + ((t ^ (t >> 7)) * (t | 61))) & 0xFFFFFFFF
        return ((t ^ (t >> 14)) & 0xFFFFFFFF) / 4294967296

    return rng


def _gauss(rng) -> float:
    u = rng()
    v = rng()
    if u < 1e-9:
        u = 1e-9
    return math.sqrt(-2 * math.log(u)) * math.cos(2 * math.pi * v)


def _log_normal_days(rng, median: float, sigma: float) -> float:
    if median <= 0:
        return 0.0
    return math.exp(math.log(median) + sigma * _gauss(rng))


def _clamp01(x: float) -> float:
    if x < 0:
        return 0.0
    if x > 1:
        return 1.0
    return x


def _seed_from_funnel(funnel: dict[str, int]) -> int:
    h = 0x811C9DC5
    for stage in PROGRESSION:
        h ^= (funnel.get(stage, 0) & 0xFF)
        h = (h * 0x01000193) & 0xFFFFFFFF
    return h or 1


def _iso_date(ms: int) -> str:
    return datetime.fromtimestamp(ms / 1000, tz=timezone.utc).strftime("%Y-%m-%d")


def _parse_iso_day(iso: str) -> int:
    dt = datetime.strptime(iso, "%Y-%m-%d").replace(tzinfo=timezone.utc)
    return int(dt.timestamp() * 1000)


# ---------- core MC ----------


def _run_mc(
    funnel: dict[str, int],
    target_ms: int,
    now: int,
    assumptions: ForecastAssumptions,
    trials: int,
    seed: int,
) -> dict:
    rng = _mulberry32(seed)
    sorted_hire_dates: list[int] = []
    by_target = 0
    hire_sum = 0
    reaches = {k: 0 for k in PROGRESSION}
    hires = {k: 0 for k in PROGRESSION}

    for _trial in range(trials):
        earliest: Optional[int] = None
        trial_hires = 0
        for stage in PROGRESSION:
            count = funnel.get(stage, 0) or 0
            for _ in range(count):
                t_cur = now
                dropped = False
                start_idx = PROGRESSION.index(stage)
                for s in range(start_idx, len(PROGRESSION)):
                    st = PROGRESSION[s]
                    reaches[st] += 1
                    days = _log_normal_days(rng, assumptions.velocity[st], assumptions.duration_sigma)
                    t_cur += days * DAY_MS
                    if rng() >= _clamp01(assumptions.conversion[st]):
                        dropped = True
                        break
                if not dropped:
                    hire_ms = t_cur + assumptions.notice_period_days * DAY_MS
                    trial_hires += 1
                    hires[stage] += 1
                    if earliest is None or hire_ms < earliest:
                        earliest = hire_ms
        if earliest is not None:
            sorted_hire_dates.append(int(earliest))
            if earliest <= target_ms:
                by_target += 1
        hire_sum += trial_hires
    sorted_hire_dates.sort()
    return {
        "any_hire": len(sorted_hire_dates) / trials,
        "by_target_prob": by_target / trials,
        "sorted_hire_dates": sorted_hire_dates,
        "expected_hires": hire_sum / trials,
        "reaches": reaches,
        "hires": hires,
    }


def _percentile(sorted_list: list[int], p: float) -> Optional[int]:
    if not sorted_list:
        return None
    idx = min(len(sorted_list) - 1, max(0, math.floor(p * len(sorted_list))))
    return sorted_list[idx]


# ---------- public API ----------


def forecast_funnel(inp: ForecastInput) -> dict:
    now = inp.now if inp.now is not None else int(time.time() * 1000)
    trials = max(200, inp.trials)
    assumptions = inp.assumptions or ForecastAssumptions()
    funnel = {k: int(inp.funnel.get(k, 0) or 0) for k in PROGRESSION}
    target_ms = _parse_iso_day(inp.target_date)
    seed = inp.seed if inp.seed is not None else _seed_from_funnel(funnel)

    total = sum(funnel.values())
    if total == 0:
        return {
            "trials": trials,
            "targetDate": inp.target_date,
            "now": now,
            "probabilityByTarget": 0.0,
            "hireDate": {"p10": None, "p50": None, "p90": None, "anyHireProbability": 0.0},
            "expectedHires": 0.0,
            "funnel": [
                {"key": k, "here": 0, "expectedAdvancers": 0.0, "expectedHires": 0.0}
                for k in PROGRESSION
            ],
            "bottleneck": None,
            "sensitivity": [],
            "recommendations": [
                "No candidates in the pipeline yet — add a shortlist before forecasting.",
            ],
            "assumptions": _assumptions_payload(assumptions),
        }

    main = _run_mc(funnel, target_ms, now, assumptions, trials, seed)

    funnel_out = [
        {
            "key": k,
            "here": funnel[k],
            "expectedAdvancers": round(main["reaches"][k] / trials, 2),
            "expectedHires": round(main["hires"][k] / trials, 2),
        }
        for k in PROGRESSION
    ]

    bottleneck = None
    best_drop = 0.0
    for k in PROGRESSION:
        reach_mean = main["reaches"][k] / trials
        drop = reach_mean * (1 - _clamp01(assumptions.conversion[k]))
        if drop > best_drop:
            best_drop = drop
            bottleneck = k

    baseline = main["by_target_prob"]
    sens_trials = max(400, trials // 4)
    sensitivity: list[dict] = []

    for stage in PROGRESSION:
        cv = assumptions.conversion[stage]
        cv_plus = _clamp01(cv + 0.15)
        cv_minus = _clamp01(cv - 0.15)

        a_plus = ForecastAssumptions(
            conversion={**assumptions.conversion, stage: cv_plus},
            velocity=dict(assumptions.velocity),
            notice_period_days=assumptions.notice_period_days,
            duration_sigma=assumptions.duration_sigma,
        )
        a_minus = ForecastAssumptions(
            conversion={**assumptions.conversion, stage: cv_minus},
            velocity=dict(assumptions.velocity),
            notice_period_days=assumptions.notice_period_days,
            duration_sigma=assumptions.duration_sigma,
        )

        plus_p = _run_mc(funnel, target_ms, now, a_plus, sens_trials, seed ^ 0x9E3779B1)["by_target_prob"]
        minus_p = _run_mc(funnel, target_ms, now, a_minus, sens_trials, seed ^ 0x85EBCA77)["by_target_prob"]
        sensitivity.append({
            "lever": {"kind": "conversion", "stage": stage},
            "label": f"{STAGE_LABEL[stage]} conversion",
            "baseline": baseline,
            "upliftPlus": plus_p,
            "upliftMinus": minus_p,
            "delta": abs(plus_p - baseline) + abs(baseline - minus_p),
        })

        v = assumptions.velocity[stage]
        a_fast = ForecastAssumptions(
            conversion=dict(assumptions.conversion),
            velocity={**assumptions.velocity, stage: max(0.25, v * 0.7)},
            notice_period_days=assumptions.notice_period_days,
            duration_sigma=assumptions.duration_sigma,
        )
        a_slow = ForecastAssumptions(
            conversion=dict(assumptions.conversion),
            velocity={**assumptions.velocity, stage: v * 1.3},
            notice_period_days=assumptions.notice_period_days,
            duration_sigma=assumptions.duration_sigma,
        )
        fast_p = _run_mc(funnel, target_ms, now, a_fast, sens_trials, seed ^ 0xC2B2AE35)["by_target_prob"]
        slow_p = _run_mc(funnel, target_ms, now, a_slow, sens_trials, seed ^ 0x27D4EB2F)["by_target_prob"]
        sensitivity.append({
            "lever": {"kind": "velocity", "stage": stage},
            "label": f"{STAGE_LABEL[stage]} speed",
            "baseline": baseline,
            "upliftPlus": fast_p,
            "upliftMinus": slow_p,
            "delta": abs(fast_p - baseline) + abs(baseline - slow_p),
        })

    for stage in ("new", "outreach"):
        bump = {**funnel, stage: funnel[stage] + 5}
        cut = {**funnel, stage: max(0, funnel[stage] - 2)}
        plus_p = _run_mc(bump, target_ms, now, assumptions, sens_trials, seed ^ 0x165667B1)["by_target_prob"]
        minus_p = _run_mc(cut, target_ms, now, assumptions, sens_trials, seed ^ 0xD3A2646C)["by_target_prob"]
        sensitivity.append({
            "lever": {"kind": "add_candidates", "stage": stage},
            "label": f"Add 5 to {STAGE_LABEL[stage]}",
            "baseline": baseline,
            "upliftPlus": plus_p,
            "upliftMinus": minus_p,
            "delta": abs(plus_p - baseline) + abs(baseline - minus_p),
        })

    sensitivity.sort(key=lambda r: r["delta"], reverse=True)

    recs: list[str] = []
    p10_ms = _percentile(main["sorted_hire_dates"], 0.10)
    p50_ms = _percentile(main["sorted_hire_dates"], 0.50)
    p90_ms = _percentile(main["sorted_hire_dates"], 0.90)

    if baseline >= 0.75:
        recs.append(
            f"Strong shape — {round(baseline * 100)}% chance to close by "
            f"{inp.target_date}. Focus on keeping the top of the funnel warm."
        )
    elif baseline >= 0.4:
        recs.append(
            f"Tight but feasible — {round(baseline * 100)}% chance to close by "
            f"{inp.target_date}. Apply the top lever below to tip it past 75%."
        )
    elif baseline > 0:
        recs.append(
            f"At risk — only {round(baseline * 100)}% chance to close by "
            f"{inp.target_date}. Consider widening the funnel or moving the date."
        )
    else:
        recs.append(
            f"Almost certain to miss — pipeline can't realistically close by "
            f"{inp.target_date}. Push the target out or escalate sourcing."
        )

    if bottleneck:
        recs.append(
            f"The {STAGE_LABEL[bottleneck]} stage is your dropout cliff — "
            "tighten the bar earlier or coach the panel to convert more of them."
        )

    if sensitivity and sensitivity[0]["delta"] > 0.05:
        top = sensitivity[0]
        arrow = "+" if top["upliftPlus"] > top["upliftMinus"] else "-"
        swing = max(
            abs(top["upliftPlus"] - top["baseline"]),
            abs(top["baseline"] - top["upliftMinus"]),
        )
        recs.append(
            f"Biggest lever: {top['label']} — pushing it favourably moves "
            f"P(hire-by-target) by {arrow}{round(swing * 100)} points."
        )

    if p50_ms is not None:
        recs.append(
            f"Median earliest-hire date: {_iso_date(p50_ms)} "
            f"(P10 {_iso_date(p10_ms) if p10_ms is not None else '—'} · "
            f"P90 {_iso_date(p90_ms) if p90_ms is not None else '—'})."
        )

    return {
        "trials": trials,
        "targetDate": inp.target_date,
        "now": now,
        "probabilityByTarget": baseline,
        "hireDate": {
            "p10": _iso_date(p10_ms) if p10_ms is not None else None,
            "p50": _iso_date(p50_ms) if p50_ms is not None else None,
            "p90": _iso_date(p90_ms) if p90_ms is not None else None,
            "anyHireProbability": main["any_hire"],
        },
        "expectedHires": round(main["expected_hires"], 2),
        "funnel": funnel_out,
        "bottleneck": bottleneck,
        "sensitivity": sensitivity,
        "recommendations": recs,
        "assumptions": _assumptions_payload(assumptions),
    }


def _assumptions_payload(a: ForecastAssumptions) -> dict:
    return {
        "conversion": dict(a.conversion),
        "velocity": dict(a.velocity),
        "noticePeriodDays": a.notice_period_days,
        "durationSigma": a.duration_sigma,
    }

File: Credicrew/frontend/src/lib/forecast.ts
// Forecast Studio engine.
//
// Every other Credicrew surface is descriptive: who's in the funnel, who
// scored what, who's offering what comp, which channel is paying off.
// None of them answer the question that every hiring manager opens with —
// **"Will I have a hire by start_date?"** Forecast Studio is the missing
// predictive layer.
//
// The engine takes the current pipeline shape (counts at each stage), the
// recruiter's conversion + velocity assumptions (default-seeded from
// industry medians, fully overridable), and runs a Monte-Carlo simulation
// of every candidate's forward walk through the funnel. Each trial yields
// an earliest hire-date (the date the strongest accepted offer hits the
// end of its notice period); aggregating thousands of trials gives:
//
//   • probability of a hire by the target date,
//   • a P10 / P50 / P90 fan chart of earliest-hire dates,
//   • the expected number of advancers reaching each downstream stage,
//   • the **bottleneck** stage (the one with the worst marginal
//     contribution to the dropout cliff),
//   • a **sensitivity tornado** ranking which lever (conversion or
//     velocity at each stage) moves P(hire-by-target) the most, and
//   • a concrete **recommendations** list ("add 4 candidates to Outreach
//     to hit Jun 20 at 75%").
//
// Pure functions. Deterministic for a given seed so the UI doesn't
// flicker on re-render. Mirrored in `backend/app/services/forecast.py`.

import type { Role } from '@/lib/roles';
import { countByStatus } from '@/lib/roles';

// ---------- types ----------

export type ForecastStage = 'new' | 'outreach' | 'screening' | 'interview' | 'offer';

export const FORECAST_PROGRESSION: ForecastStage[] = [
  'new',
  'outreach',
  'screening',
  'interview',
  'offer',
];

export const FORECAST_STAGE_LABEL: Record<ForecastStage, string> = {
  new: 'New',
  outreach: 'Outreach',
  screening: 'Screening',
  interview: 'Interview',
  offer: 'Offer',
};

/** Probability of advancing from this stage to the next. 0..1 */
export type ConversionMap = Record<ForecastStage, number>;

/** Median days spent at this stage before advancing or dropping. */
export type VelocityMap = Record<ForecastStage, number>;

export type ForecastAssumptions = {
  /** p(advance from stage → next). offer → "accept". */
  conversion: ConversionMap;
  /** median days at each stage (LogNormal median). offer → days to decision. */
  velocity: VelocityMap;
  /** post-accept notice period before the candidate can start. */
  noticePeriodDays: number;
  /** LogNormal sigma — higher = more variance per stage. */
  durationSigma: number;
};

export const DEFAULT_CONVERSION: ConversionMap = {
  new: 0.65,
  outreach: 0.40,
  screening: 0.60,
  interview: 0.35,
  offer: 0.70,
};

export const DEFAULT_VELOCITY: VelocityMap = {
  new: 2,
  outreach: 4,
  screening: 5,
  interview: 8,
  offer: 4,
};

export const DEFAULT_ASSUMPTIONS: ForecastAssumptions = {
  conversion: DEFAULT_CONVERSION,
  velocity: DEFAULT_VELOCITY,
  noticePeriodDays: 30,
  durationSigma: 0.45,
};

export type ForecastInput = {
  /** counts of candidates currently sitting at each stage. */
  funnel: Record<ForecastStage, number>;
  /** ISO date string (YYYY-MM-DD) the recruiter is targeting for start. */
  targetDate: string;
  /** ms since epoch — the "now" anchor for the simulation. */
  now?: number;
  assumptions?: Partial<ForecastAssumptions>;
  /** Monte-Carlo trial count. Defaults to 4000. */
  trials?: number;
  /** RNG seed. Defaults to a stable value of the funnel. */
  seed?: number;
};

export type ForecastFunnelStage = {
  key: ForecastStage;
  here: number;
  expectedAdvancers: number; // E[# who reach the next stage]
  expectedHires: number;     // E[# who eventually become hires]
};

export type ForecastSensitivityRow = {
  lever:
    | { kind: 'conversion'; stage: ForecastStage }
    | { kind: 'velocity'; stage: ForecastStage }
    | { kind: 'add_candidates'; stage: ForecastStage };
  label: string;
  baseline: number;          // P(hire-by-target) baseline
  upliftPlus: number;        // P after positive nudge
  upliftMinus: number;       // P after negative nudge
  delta: number;             // |+| + |-| — total swing
};

export type ForecastResult = {
  trials: number;
  targetDate: string;
  now: number;
  /** P(at least one hire on/before target date). 0..1 */
  probabilityByTarget: number;
  /** Distribution of earliest-hire date over trials with any hire. */
  hireDate: {
    p10: string | null;
    p50: string | null;
    p90: string | null;
    /** fraction of trials that produced ANY hire (any date). */
    anyHireProbability: number;
  };
  /** Mean # of hires per trial. */
  expectedHires: number;
  funnel: ForecastFunnelStage[];
  /** Stage whose marginal probability contributes most to the dropout cliff. */
  bottleneck: ForecastStage | null;
  /** Ranked levers by total swing in P(hire-by-target). */
  sensitivity: ForecastSensitivityRow[];
  /** Pithy action suggestions. */
  recommendations: string[];
  assumptions: ForecastAssumptions;
};

// ---------- utils ----------

const DAY_MS = 86_400_000;

/** Deterministic mulberry32 PRNG. */
function mulberry32(seed: number): () => number {
  let a = seed >>> 0;
  return function () {
    a = (a + 0x6D2B79F5) >>> 0;
    let t = a;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

/** Box-Muller transform for a standard normal sample. */
function gauss(rng: () => number): number {
  let u = rng();
  const v = rng();
  if (u < 1e-9) u = 1e-9;
  return Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v);
}

/** LogNormal sample with the given median and sigma. */
function logNormalDays(rng: () => number, median: number, sigma: number): number {
  if (median <= 0) return 0;
  return Math.exp(Math.log(median) + sigma * gauss(rng));
}

function clamp01(x: number): number {
  return x < 0 ? 0 : x > 1 ? 1 : x;
}

function isoDate(ms: number): string {
  return new Date(ms).toISOString().slice(0, 10);
}

function parseIsoDay(iso: string): number {
  // Pin to start of UTC day so target comparisons are inclusive.
  return Date.parse(`${iso}T00:00:00.000Z`);
}

function mergeAssumptions(over?: Partial<ForecastAssumptions>): ForecastAssumptions {
  if (!over) return DEFAULT_ASSUMPTIONS;
  return {
    conversion: { ...DEFAULT_CONVERSION, ...(over.conversion ?? {}) },
    velocity: { ...DEFAULT_VELOCITY, ...(over.velocity ?? {}) },
    noticePeriodDays: over.noticePeriodDays ?? DEFAULT_ASSUMPTIONS.noticePeriodDays,
    durationSigma: over.durationSigma ?? DEFAULT_ASSUMPTIONS.durationSigma,
  };
}

function seedFromFunnel(funnel: Record<ForecastStage, number>): number {
  // FNV-1a over the funnel counts so re-rendering with the same shape
  // returns identical numbers but reshuffles when candidates move.
  let h = 0x811c9dc5;
  for (const stage of FORECAST_PROGRESSION) {
    h ^= funnel[stage] & 0xff;
    h = (h * 0x01000193) >>> 0;
  }
  return (h || 1) >>> 0;
}

// ---------- engine ----------

/** Build an input funnel from a saved Role. */
export function funnelFromRole(role: Role): Record<ForecastStage, number> {
  const counts = countByStatus(role);
  return {
    new: counts.new ?? 0,
    outreach: counts.outreach ?? 0,
    screening: counts.screening ?? 0,
    interview: counts.interview ?? 0,
    offer: counts.offer ?? 0,
  };
}

/**
 * Core Monte-Carlo runner. Returns:
 *   anyHire: fraction of trials with at least one hire,
 *   hits: trials that produced a hire by `targetMs`,
 *   sortedHireDates: sorted earliest-hire-date per trial (ms),
 *   expectedHires: mean # of hires per trial,
 *   reaches: per-stage reach counts summed across trials.
 */
function runMc(
  funnel: Record<ForecastStage, number>,
  targetMs: number,
  now: number,
  a: ForecastAssumptions,
  trials: number,
  seed: number,
): {
  anyHire: number;
  byTargetProb: number;
  sortedHireDates: number[];
  expectedHires: number;
  reaches: Record<ForecastStage, number>;
  hires: Record<ForecastStage, number>; // # who started at stage S and hired
} {
  const rng = mulberry32(seed);
  const sortedHireDates: number[] = [];
  let byTarget = 0;
  let hireSum = 0;
  const reaches: Record<ForecastStage, number> = {
    new: 0, outreach: 0, screening: 0, interview: 0, offer: 0,
  };
  const hires: Record<ForecastStage, number> = {
    new: 0, outreach: 0, screening: 0, interview: 0, offer: 0,
  };

  for (let t = 0; t < trials; t++) {
    let earliest: number | null = null;
    let trialHires = 0;
    for (const stage of FORECAST_PROGRESSION) {
      const count = funnel[stage] | 0;
      for (let i = 0; i < count; i++) {
        // Simulate stage walk and count reach side-effect inline.
        let tcur = now;
        let dropped = false;
        const startIdx = FORECAST_PROGRESSION.indexOf(stage);
        for (let s = startIdx; s < FORECAST_PROGRESSION.length; s++) {
          const st = FORECAST_PROGRESSION[s];
          reaches[st] += 1;
          const days = logNormalDays(rng, a.velocity[st], a.durationSigma);
          tcur += days * DAY_MS;
          if (rng() >= clamp01(a.conversion[st])) {
            dropped = true;
            break;
          }
        }
        if (!dropped) {
          const hireMs = tcur + a.noticePeriodDays * DAY_MS;
          trialHires += 1;
          hires[stage] += 1;
          if (earliest === null || hireMs < earliest) earliest = hireMs;
        }
      }
    }
    if (earliest !== null) {
      sortedHireDates.push(earliest);
      if (earliest <= targetMs) byTarget += 1;
    }
    hireSum += trialHires;
  }
  sortedHireDates.sort((a, b) => a - b);
  return {
    anyHire: sortedHireDates.length / trials,
    byTargetProb: byTarget / trials,
    sortedHireDates,
    expectedHires: hireSum / trials,
    reaches,
    hires,
  };
}

function percentile(sorted: number[], p: number): number | null {
  if (sorted.length === 0) return null;
  const idx = Math.min(sorted.length - 1, Math.max(0, Math.floor(p * sorted.length)));
  return sorted[idx];
}

/** A quick deterministic baseline — no MC needed — for sensitivity speedups. */
function fastByTargetEstimate(
  funnel: Record<ForecastStage, number>,
  targetMs: number,
  now: number,
  a: ForecastAssumptions,
): number {
  // Expected # of hires assuming everyone takes ~median time. We use a
  // closed-form lower-bound: 1 - P(zero hires by target).
  let probZero = 1;
  for (const stage of FORECAST_PROGRESSION) {
    const count = funnel[stage] | 0;
    if (count === 0) continue;
    const startIdx = FORECAST_PROGRESSION.indexOf(stage);
    // p that this candidate succeeds AND finishes by target
    let pSuccess = 1;
    let cumDays = 0;
    for (let s = startIdx; s < FORECAST_PROGRESSION.length; s++) {
      const st = FORECAST_PROGRESSION[s];
      pSuccess *= clamp01(a.conversion[st]);
      cumDays += a.velocity[st];
    }
    cumDays += a.noticePeriodDays;
    const finishMs = now + cumDays * DAY_MS;
    if (finishMs > targetMs) {
      // Apply a log-normal CDF-ish heuristic — for the fast pass we just
      // pin to 0 if the median path overruns; the full MC catches the rest.
      pSuccess *= 0.25;
    }
    probZero *= Math.pow(1 - pSuccess, count);
  }
  return 1 - probZero;
}

// ---------- public API ----------

/** Run the forecast. */
export function forecastFunnel(input: ForecastInput): ForecastResult {
  const now = input.now ?? Date.now();
  const trials = Math.max(200, input.trials ?? 4000);
  const assumptions = mergeAssumptions(input.assumptions);
  const funnel = input.funnel;
  const targetMs = parseIsoDay(input.targetDate);
  const seed = input.seed ?? seedFromFunnel(funnel);

  const total = FORECAST_PROGRESSION.reduce((s, k) => s + (funnel[k] | 0), 0);

  // Empty pipeline → nothing to forecast.
  if (total === 0) {
    return {
      trials,
      targetDate: input.targetDate,
      now,
      probabilityByTarget: 0,
      hireDate: { p10: null, p50: null, p90: null, anyHireProbability: 0 },
      expectedHires: 0,
      funnel: FORECAST_PROGRESSION.map(k => ({
        key: k,
        here: 0,
        expectedAdvancers: 0,
        expectedHires: 0,
      })),
      bottleneck: null,
      sensitivity: [],
      recommendations: [
        'No candidates in the pipeline yet — add a shortlist before forecasting.',
      ],
      assumptions,
    };
  }

  const main = runMc(funnel, targetMs, now, assumptions, trials, seed);

  const funnelOut: ForecastFunnelStage[] = FORECAST_PROGRESSION.map(k => ({
    key: k,
    here: funnel[k] | 0,
    expectedAdvancers: +(main.reaches[k] / trials).toFixed(2),
    expectedHires: +(main.hires[k] / trials).toFixed(2),
  }));

  // Bottleneck: stage with the biggest marginal probability mass lost.
  // We compute, for each stage, the *upstream survivors expected at that
  // stage* times (1 - conversion). The stage with the largest such drop
  // is where the most candidates die.
  const expectedReachAt: Record<ForecastStage, number> = { ...main.reaches };
  let bottleneck: ForecastStage | null = null;
  let bestDrop = 0;
  for (const k of FORECAST_PROGRESSION) {
    const reachMean = expectedReachAt[k] / trials;
    const drop = reachMean * (1 - clamp01(assumptions.conversion[k]));
    if (drop > bestDrop) {
      bestDrop = drop;
      bottleneck = k;
    }
  }

  // Sensitivity — tornado over conversion + velocity at each stage.
  // We use a smaller MC budget for each perturbation to keep things snappy.
  const sensTrials = Math.max(400, Math.floor(trials / 4));
  const baseline = main.byTargetProb;

  const sensitivity: ForecastSensitivityRow[] = [];

  for (const stage of FORECAST_PROGRESSION) {
    // Conversion ±15 percentage points (clamped to [0,1]).
    const cv = assumptions.conversion[stage];
    const cvPlus = clamp01(cv + 0.15);
    const cvMinus = clamp01(cv - 0.15);
    const aPlus = {
      ...assumptions,
      conversion: { ...assumptions.conversion, [stage]: cvPlus },
    };
    const aMinus = {
      ...assumptions,
      conversion: { ...assumptions.conversion, [stage]: cvMinus },
    };
    const plusP = runMc(funnel, targetMs, now, aPlus, sensTrials, seed ^ 0x9e3779b1).byTargetProb;
    const minusP = runMc(funnel, targetMs, now, aMinus, sensTrials, seed ^ 0x85ebca77).byTargetProb;
    sensitivity.push({
      lever: { kind: 'conversion', stage },
      label: `${FORECAST_STAGE_LABEL[stage]} conversion`,
      baseline,
      upliftPlus: plusP,
      upliftMinus: minusP,
      delta: Math.abs(plusP - baseline) + Math.abs(baseline - minusP),
    });

    // Velocity ±30%.
    const v = assumptions.velocity[stage];
    const aFast = {
      ...assumptions,
      velocity: { ...assumptions.velocity, [stage]: Math.max(0.25, v * 0.7) },
    };
    const aSlow = {
      ...assumptions,
      velocity: { ...assumptions.velocity, [stage]: v * 1.3 },
    };
    const fastP = runMc(funnel, targetMs, now, aFast, sensTrials, seed ^ 0xc2b2ae35).byTargetProb;
    const slowP = runMc(funnel, targetMs, now, aSlow, sensTrials, seed ^ 0x27d4eb2f).byTargetProb;
    sensitivity.push({
      lever: { kind: 'velocity', stage },
      label: `${FORECAST_STAGE_LABEL[stage]} speed`,
      baseline,
      upliftPlus: fastP,
      upliftMinus: slowP,
      delta: Math.abs(fastP - baseline) + Math.abs(baseline - slowP),
    });
  }

  // Add-candidates lever — only useful for the upstream stages where the
  // recruiter can plausibly inject more pipeline mass.
  for (const stage of ['new', 'outreach'] as ForecastStage[]) {
    const bumpFunnel = { ...funnel, [stage]: (funnel[stage] | 0) + 5 };
    const cutFunnel = { ...funnel, [stage]: Math.max(0, (funnel[stage] | 0) - 2) };
    const plusP = runMc(bumpFunnel, targetMs, now, assumptions, sensTrials, seed ^ 0x165667b1).byTargetProb;
    const minusP = runMc(cutFunnel, targetMs, now, assumptions, sensTrials, seed ^ 0xd3a2646c).byTargetProb;
    sensitivity.push({
      lever: { kind: 'add_candidates', stage },
      label: `Add 5 to ${FORECAST_STAGE_LABEL[stage]}`,
      baseline,
      upliftPlus: plusP,
      upliftMinus: minusP,
      delta: Math.abs(plusP - baseline) + Math.abs(baseline - minusP),
    });
  }

  sensitivity.sort((a, b) => b.delta - a.delta);

  // Recommendations — pull from the top sensitivity rows + funnel facts.
  const recs: string[] = [];
  const p10ms = percentile(main.sortedHireDates, 0.10);
  const p50ms = percentile(main.sortedHireDates, 0.50);
  const p90ms = percentile(main.sortedHireDates, 0.90);

  if (baseline >= 0.75) {
    recs.push(
      `Strong shape — ${Math.round(baseline * 100)}% chance to close by ${input.targetDate}. Focus on keeping the top of the funnel warm.`,
    );
  } else if (baseline >= 0.4) {
    recs.push(
      `Tight but feasible — ${Math.round(baseline * 100)}% chance to close by ${input.targetDate}. Apply the top lever below to tip it past 75%.`,
    );
  } else if (baseline > 0) {
    recs.push(
      `At risk — only ${Math.round(baseline * 100)}% chance to close by ${input.targetDate}. Consider widening the funnel or moving the date.`,
    );
  } else {
    recs.push(
      `Almost certain to miss — pipeline can't realistically close by ${input.targetDate}. Push the target out or escalate sourcing.`,
    );
  }

  if (bottleneck) {
    recs.push(
      `The ${FORECAST_STAGE_LABEL[bottleneck]} stage is your dropout cliff — ` +
        `tighten the bar earlier or coach the panel to convert more of them.`,
    );
  }
  const top = sensitivity[0];
  if (top && top.delta > 0.05) {
    const arrow = top.upliftPlus > top.upliftMinus ? '+' : '-';
    recs.push(
      `Biggest lever: ${top.label} — pushing it favourably moves P(hire-by-target) by ${arrow}${Math.round(
        Math.max(Math.abs(top.upliftPlus - top.baseline), Math.abs(top.baseline - top.upliftMinus)) * 100,
      )} points.`,
    );
  }
  if (p50ms !== null) {
    recs.push(
      `Median earliest-hire date: ${isoDate(p50ms)} (P10 ${
        p10ms !== null ? isoDate(p10ms) : '—'
      } · P90 ${p90ms !== null ? isoDate(p90ms) : '—'}).`,
    );
  }

  return {
    trials,
    targetDate: input.targetDate,
    now,
    probabilityByTarget: baseline,
    hireDate: {
      p10: p10ms !== null ? isoDate(p10ms) : null,
      p50: p50ms !== null ? isoDate(p50ms) : null,
      p90: p90ms !== null ? isoDate(p90ms) : null,
      anyHireProbability: main.anyHire,
    },
    expectedHires: +main.expectedHires.toFixed(2),
    funnel: funnelOut,
    bottleneck,
    sensitivity,
    recommendations: recs,
    assumptions,
  };
}

/**
 * Closed-form fast estimate of P(hire-by-target). Used by the UI for
 * the "quick what-if" sliders that update on drag without re-running
 * Monte Carlo.
 */
export function quickProbabilityEstimate(
  funnel: Record<ForecastStage, number>,
  targetDate: string,
  now: number,
  overrides?: Partial<ForecastAssumptions>,
): number {
  return fastByTargetEstimate(
    funnel,
    parseIsoDay(targetDate),
    now,
    mergeAssumptions(overrides),
  );
}

/** Default target date: 60 days from now. */
export function defaultTargetDate(now: number = Date.now()): string {
  return isoDate(now + 60 * DAY_MS);
}
